File: src/pages/api/steepness.ts
// Section F4: Curve Steepness (10Y − 2Y)
// Germany removed: no daily 2Y data available via any free API.
// Norway 2Y sourced from GOVT_ZEROCOUPON (daily).
// US 2Y and 10Y from FRED (daily).

import type { NextApiRequest, NextApiResponse } from "next";
import { fredSeries, nDaysAgo } from "@/lib/fred";
import { nbFetchYield } from "@/lib/norgesbank";
import { getCache, setCache } from "@/lib/cache";
import { alignByDate } from "@/lib/stats";

const CACHE_KEY = "steepness";

export interface SteepnessEntry {
  country: string;
  current: number | null;
  weekAgo: number | null;
  direction: "steeper" | "flatter" | "unchanged" | "unavailable";
  inverted: boolean;
}

export interface SteepnessResponse {
  entries: SteepnessEntry[];
  updatedAt: string;
}

function direction(current: number | null, weekAgo: number | null): SteepnessEntry["direction"] {
  if (current === null || weekAgo === null) return "unavailable";
  const diff = current - weekAgo;
  if (Math.abs(diff) < 0.01) return "unchanged";
  return diff > 0 ? "steeper" : "flatter";
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<SteepnessResponse | { error: string }>
) {
  const cached = getCache<SteepnessResponse>(CACHE_KEY);
  if (cached) return res.status(200).json(cached);

  // ~14 trading days for the "last week" comparison
  const start = nDaysAgo(22);

  try {
    const [us10Raw, us2Raw, no10Raw, no2Raw] = await Promise.all([
      fredSeries("DGS10", start).catch(() => []),
      fredSeries("DGS2",  start).catch(() => []),
      nbFetchYield("10Y", start).catch(() => []),
      nbFetchYield("2Y",  start).catch(() => []),  // GOVT_ZEROCOUPON
    ]);

    const us10 = us10Raw.map((p) => ({ date: p.date, value: parseFloat(p.value) }));
    const us2  = us2Raw.map((p)  => ({ date: p.date, value: parseFloat(p.value) }));
    const no10 = no10Raw.map((p) => ({ date: p.date, value: p.value }));
    const no2  = no2Raw.map((p)  => ({ date: p.date, value: p.value }));

    // US
    const usAligned = alignByDate(us10, us2);
    const usCurrent = usAligned.length ? usAligned[usAligned.length - 1].a - usAligned[usAligned.length - 1].b : null;
    const usWeekAgo = usAligned.length > 5 ? usAligned[usAligned.length - 6].a - usAligned[usAligned.length - 6].b : null;

    // Norway — both from daily sources
    const noAligned = alignByDate(no10, no2);
    const noCurrent = noAligned.length ? noAligned[noAligned.length - 1].a - noAligned[noAligned.length - 1].b : null;
    const noWeekAgo = noAligned.length > 5 ? noAligned[noAligned.length - 6].a - noAligned[noAligned.length - 6].b : null;

    const round2 = (v: number | null) => v !== null ? parseFloat(v.toFixed(2)) : null;

    const entries: SteepnessEntry[] = [
      {
        country: "Norway",
        current: round2(noCurrent),
        weekAgo: round2(noWeekAgo),
        direction: direction(noCurrent, noWeekAgo),
        inverted: noCurrent !== null && noCurrent < 0,
      },
      {
        country: "United States",
        current: round2(usCurrent),
        weekAgo: round2(usWeekAgo),
        direction: direction(usCurrent, usWeekAgo),
        inverted: usCurrent !== null && usCurrent < 0,
      },
    ];

    const result: SteepnessResponse = { entries, updatedAt: new Date().toISOString() };
    setCache(CACHE_KEY, result);
    res.status(200).json(result);
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
}

File: backend/routers/steepness.py
from datetime import datetime
from typing import Optional

import asyncio
from fastapi import APIRouter
from fastapi.responses import JSONResponse

from lib.cache import get_cache, set_cache
from lib.fred import fred_series, n_days_ago
from lib.norgesbank import nb_fetch_yield
from lib.stats import align_by_date

router = APIRouter()
CACHE_KEY = "steepness"


def _direction(current: Optional[float], week_ago: Optional[float]) -> str:
    if current is None or week_ago is None:
        return "unavailable"
    diff = current - week_ago
    if abs(diff) < 0.01:
        return "unchanged"
    return "steeper" if diff > 0 else "flatter"


def _round2(v: Optional[float]) -> Optional[float]:
    return round(v, 2) if v is not None else None


@router.get("/steepness")
async def steepness():
    cached = get_cache(CACHE_KEY)
    if cached:
        return cached

    start = n_days_ago(22)

    try:
        us10_raw, us2_raw, no10_raw, no2_raw = await asyncio.gather(
            fred_series("DGS10", start),
            fred_series("DGS2",  start),
            nb_fetch_yield("10Y", start),
            nb_fetch_yield("2Y",  start),   # GOVT_ZEROCOUPON
            return_exceptions=True,
        )
        if isinstance(us10_raw, Exception): us10_raw = []
        if isinstance(us2_raw, Exception):  us2_raw = []
        if isinstance(no10_raw, Exception): no10_raw = []
        if isinstance(no2_raw, Exception):  no2_raw = []

        us_aligned = align_by_date(us10_raw, us2_raw)
        no_aligned = align_by_date(no10_raw, no2_raw)

        us_current = (us_aligned[-1]["a"] - us_aligned[-1]["b"]) if us_aligned else None
        us_week_ago = (us_aligned[-6]["a"] - us_aligned[-6]["b"]) if len(us_aligned) > 5 else None

        no_current = (no_aligned[-1]["a"] - no_aligned[-1]["b"]) if no_aligned else None
        no_week_ago = (no_aligned[-6]["a"] - no_aligned[-6]["b"]) if len(no_aligned) > 5 else None

        entries = [
            {
                "country": "Norway",
                "current": _round2(no_current),
                "weekAgo": _round2(no_week_ago),
                "direction": _direction(no_current, no_week_ago),
                "inverted": (no_current < 0) if no_current is not None else False,
            },
            {
                "country": "United States",
                "current": _round2(us_current),
                "weekAgo": _round2(us_week_ago),
                "direction": _direction(us_current, us_week_ago),
                "inverted": (us_current < 0) if us_current is not None else False,
            },
        ]

        result = {"entries": entries, "updatedAt": datetime.utcnow().isoformat() + "Z"}
        set_cache(CACHE_KEY, result)
        return result
    except Exception as e:
        return JSONResponse(status_code=500, content={"error": str(e)})
